File: services/api-gateway/app/core/sentry.py
"""Sentry error tracking and performance monitoring.

Provides centralized error tracking with:
- Automatic exception capture
- Performance tracing for requests
- Custom context for debugging
- Privacy-safe scrubbing of sensitive data
"""

from typing import Any

from app.core.config import settings
from app.core.logging import get_logger

logger = get_logger(__name__)

# Track initialization state
_sentry_initialized = False


def init_sentry() -> bool:
    """Initialize Sentry SDK if DSN is configured.

    Returns:
        True if Sentry was initialized, False otherwise
    """
    global _sentry_initialized

    if _sentry_initialized:
        return True

    if not settings.SENTRY_DSN:
        logger.info("sentry_disabled", reason="SENTRY_DSN not configured")
        return False

    try:
        import sentry_sdk
        from sentry_sdk.integrations.fastapi import FastApiIntegration
        from sentry_sdk.integrations.logging import LoggingIntegration
        from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration

        sentry_sdk.init(
            dsn=settings.SENTRY_DSN,
            environment=settings.ENVIRONMENT,
            release=f"voiceassist-api@{settings.APP_VERSION}",
            traces_sample_rate=settings.SENTRY_TRACES_SAMPLE_RATE,
            profiles_sample_rate=settings.SENTRY_PROFILES_SAMPLE_RATE,
            integrations=[
                FastApiIntegration(transaction_style="endpoint"),
                SqlalchemyIntegration(),
                LoggingIntegration(
                    level=None,  # Capture all breadcrumbs
                    event_level=40,  # Only send ERROR and above as events
                ),
            ],
            # Privacy: scrub sensitive data
            send_default_pii=False,
            before_send=_scrub_sensitive_data,
            before_send_transaction=_scrub_transaction_data,
        )

        _sentry_initialized = True
        logger.info(
            "sentry_initialized",
            environment=settings.ENVIRONMENT,
            traces_sample_rate=settings.SENTRY_TRACES_SAMPLE_RATE,
        )
        return True

    except ImportError:
        logger.warning("sentry_import_error", reason="sentry-sdk not installed")
        return False
    except Exception as e:
        logger.error("sentry_init_error", error=str(e))
        return False


def _scrub_sensitive_data(event: dict, hint: dict) -> dict | None:
    """Scrub sensitive data before sending to Sentry.

    Removes:
    - Authorization headers
    - API keys
    - Passwords
    - PHI markers
    """
    if "request" in event:
        request = event["request"]

        # Scrub headers
        if "headers" in request:
            headers = request["headers"]
            sensitive_headers = ["authorization", "x-api-key", "cookie", "set-cookie"]
            for header in sensitive_headers:
                if header in headers:
                    headers[header] = "[REDACTED]"

        # Scrub query strings
        if "query_string" in request:
            request["query_string"] = _redact_params(request["query_string"])

        # Scrub body data
        if "data" in request and isinstance(request["data"], dict):
            request["data"] = _redact_dict(request["data"])

    # Scrub extra context
    if "extra" in event:
        event["extra"] = _redact_dict(event["extra"])

    return event


def _scrub_transaction_data(event: dict, hint: dict) -> dict | None:
    """Scrub sensitive data from transaction events."""
    # Remove user data from transactions
    if "user" in event:
        user = event["user"]
        # Keep only non-PII identifiers
        event["user"] = {"id": user.get("id")}

    return event


def _redact_params(params: str) -> str:
    """Redact sensitive query parameters."""
    sensitive_params = ["token", "api_key", "password", "secret", "key"]
    parts = params.split("&")
    result = []
    for part in parts:
        if "=" in part:
            key, _ = part.split("=", 1)
            if any(s in key.lower() for s in sensitive_params):
                result.append(f"{key}=[REDACTED]")
            else:
                result.append(part)
        else:
            result.append(part)
    return "&".join(result)


def _redact_dict(data: dict) -> dict:
    """Recursively redact sensitive dictionary values."""
    sensitive_keys = [
        "password",
        "token",
        "api_key",
        "secret",
        "authorization",
        "transcript",
        "content",
        "message",
    ]
    result = {}
    for key, value in data.items():
        if any(s in key.lower() for s in sensitive_keys):
            result[key] = "[REDACTED]"
        elif isinstance(value, dict):
            result[key] = _redact_dict(value)
        elif isinstance(value, list):
            result[key] = [_redact_dict(v) if isinstance(v, dict) else v for v in value]
        else:
            result[key] = value
    return result


def capture_voice_error(
    error: Exception,
    *,
    user_id: str | None = None,
    conversation_id: str | None = None,
    voice_status: str | None = None,
    extra: dict[str, Any] | None = None,
) -> str | None:
    """Capture a voice-related error to Sentry with context.

    Args:
        error: The exception to capture
        user_id: User identifier (no PII)
        conversation_id: Conversation identifier
        voice_status: Current voice session status
        extra: Additional context (will be scrubbed)

    Returns:
        Sentry event ID if captured, None otherwise
    """
    if not _sentry_initialized:
        return None

    try:
        import sentry_sdk

        with sentry_sdk.push_scope() as scope:
            scope.set_tag("feature", "voice")
            scope.set_tag("voice_status", voice_status or "unknown")

            if user_id:
                scope.set_user({"id": user_id})

            if conversation_id:
                scope.set_tag("conversation_id", conversation_id)

            if extra:
                # Scrub extra data before adding
                scrubbed = _redact_dict(extra)
                for key, value in scrubbed.items():
                    scope.set_extra(key, value)

            return sentry_sdk.capture_exception(error)

    except Exception as e:
        logger.warning("sentry_capture_error", error=str(e))
        return None


def capture_slo_violation(
    metric_name: str,
    actual_value: float,
    threshold: float,
    *,
    user_id: str | None = None,
    conversation_id: str | None = None,
) -> str | None:
    """Capture an SLO violation as a Sentry event.

    Args:
        metric_name: Name of the SLO metric
        actual_value: The measured value
        threshold: The SLO threshold that was violated
        user_id: User identifier
        conversation_id: Conversation identifier

    Returns:
        Sentry event ID if captured, None otherwise
    """
    if not _sentry_initialized:
        return None

    try:
        import sentry_sdk

        with sentry_sdk.push_scope() as scope:
            scope.set_tag("type", "slo_violation")
            scope.set_tag("metric", metric_name)
            scope.set_level("warning")

            if user_id:
                scope.set_user({"id": user_id})

            if conversation_id:
                scope.set_tag("conversation_id", conversation_id)

            scope.set_extra("actual_value", actual_value)
            scope.set_extra("threshold", threshold)
            scope.set_extra("exceeded_by", actual_value - threshold)

            msg = f"SLO violation: {metric_name} "
            msg += f"({actual_value:.0f}ms > {threshold:.0f}ms)"
            return sentry_sdk.capture_message(msg)

    except Exception as e:
        logger.warning("sentry_slo_capture_error", error=str(e))
        return None

File: apps/web-app/src/lib/sentry.ts
/**
 * Sentry Error Tracking for Frontend
 *
 * Provides:
 * - Automatic error capture
 * - Performance monitoring
 * - Privacy-safe scrubbing
 * - Voice-specific error context
 */

import * as Sentry from "@sentry/browser";

// Track initialization state
let sentryInitialized = false;

/**
 * Initialize Sentry if DSN is configured
 */
export function initSentry(): boolean {
  if (sentryInitialized) {
    return true;
  }

  const dsn = import.meta.env.VITE_SENTRY_DSN;

  if (!dsn) {
    console.log("[Sentry] Disabled: VITE_SENTRY_DSN not configured");
    return false;
  }

  try {
    Sentry.init({
      dsn,
      environment: import.meta.env.MODE,
      release: `voiceassist-web@${import.meta.env.VITE_APP_VERSION || "0.1.0"}`,

      // Performance monitoring
      tracesSampleRate: import.meta.env.PROD ? 0.1 : 1.0,

      // Privacy: don't send PII
      sendDefaultPii: false,

      // Scrub sensitive data
      beforeSend(event) {
        return scrubSensitiveData(event);
      },

      // Scrub breadcrumbs
      beforeBreadcrumb(breadcrumb) {
        // Don't log user input
        if (breadcrumb.category === "ui.input") {
          return null;
        }
        // Scrub URLs with tokens
        if (breadcrumb.data?.url) {
          breadcrumb.data.url = scrubUrl(breadcrumb.data.url);
        }
        return breadcrumb;
      },

      // Ignore common non-actionable errors
      ignoreErrors: [
        // Browser extensions
        /extensions\//i,
        /^chrome-extension:\/\//i,
        /^moz-extension:\/\//i,
        // Network errors that user can't control
        "Network request failed",
        "Failed to fetch",
        "Load failed",
        // User aborted
        "AbortError",
        "The operation was aborted",
      ],
    });

    sentryInitialized = true;
    console.log("[Sentry] Initialized", { environment: import.meta.env.MODE });
    return true;
  } catch (err) {
    console.warn("[Sentry] Failed to initialize:", err);
    return false;
  }
}

/**
 * Scrub sensitive data from events
 */
function scrubSensitiveData(
  event: Sentry.ErrorEvent,
): Sentry.ErrorEvent | null {
  // Scrub request data
  if (event.request) {
    if (event.request.headers) {
      const sensitiveHeaders = [
        "authorization",
        "cookie",
        "x-api-key",
        "set-cookie",
      ];
      for (const header of sensitiveHeaders) {
        if (event.request.headers[header]) {
          event.request.headers[header] = "[REDACTED]";
        }
      }
    }

    if (event.request.url) {
      event.request.url = scrubUrl(event.request.url);
    }
  }

  // Scrub extra context
  if (event.extra) {
    event.extra = scrubObject(event.extra);
  }

  return event;
}

/**
 * Scrub sensitive query parameters from URL
 */
function scrubUrl(url: string): string {
  try {
    const parsed = new URL(url, window.location.origin);
    const sensitiveParams = ["token", "api_key", "key", "password", "secret"];

    for (const param of sensitiveParams) {
      if (parsed.searchParams.has(param)) {
        parsed.searchParams.set(param, "[REDACTED]");
      }
    }

    return parsed.toString();
  } catch {
    return url;
  }
}

/**
 * Recursively scrub sensitive values from object
 */
function scrubObject(obj: Record<string, unknown>): Record<string, unknown> {
  const sensitiveKeys = [
    "password",
    "token",
    "api_key",
    "secret",
    "authorization",
    "transcript",
    "content",
    "message",
  ];

  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (sensitiveKeys.some((s) => key.toLowerCase().includes(s))) {
      result[key] = "[REDACTED]";
    } else if (value && typeof value === "object" && !Array.isArray(value)) {
      result[key] = scrubObject(value as Record<string, unknown>);
    } else {
      result[key] = value;
    }
  }

  return result;
}

/**
 * Capture a voice-related error with context
 */
export function captureVoiceError(
  error: Error | unknown,
  context?: {
    status?: string;
    conversationId?: string;
    metrics?: Record<string, number | null>;
    breadcrumb?: string;
  },
): string | undefined {
  if (!sentryInitialized) {
    return undefined;
  }

  const breadcrumb = context?.breadcrumb
    ? {
        type: "info" as const,
        category: "voice",
        message: context.breadcrumb,
        level: "info" as const,
      }
    : undefined;

  if (breadcrumb) {
    Sentry.addBreadcrumb(breadcrumb);
  }

  return Sentry.captureException(error, {
    tags: {
      feature: "voice",
      voice_status: context?.status || "unknown",
    },
    extra: {
      conversation_id: context?.conversationId,
      breadcrumb: context?.breadcrumb,
      // Only include numeric metrics, no transcripts
      metrics: context?.metrics
        ? {
            connectionTimeMs: context.metrics.connectionTimeMs,
            lastSttLatencyMs: context.metrics.lastSttLatencyMs,
            lastResponseLatencyMs: context.metrics.lastResponseLatencyMs,
            sessionDurationMs: context.metrics.sessionDurationMs,
          }
        : undefined,
    },
  });
}

/**
 * Capture a voice SLO violation
 */
export function captureVoiceSLOViolation(
  metric: string,
  actualMs: number,
  thresholdMs: number,
  conversationId?: string,
): string | undefined {
  if (!sentryInitialized) {
    return undefined;
  }

  return Sentry.captureMessage(
    `Voice SLO violation: ${metric} (${actualMs.toFixed(0)}ms > ${thresholdMs}ms)`,
    {
      level: "warning",
      tags: {
        feature: "voice",
        type: "slo_violation",
        metric,
      },
      extra: {
        actual_ms: actualMs,
        threshold_ms: thresholdMs,
        exceeded_by_ms: actualMs - thresholdMs,
        conversation_id: conversationId,
      },
    },
  );
}

/**
 * Set user context (privacy-safe)
 */
export function setUser(userId: string | null): void {
  if (!sentryInitialized) {
    return;
  }

  if (userId) {
    Sentry.setUser({ id: userId });
  } else {
    Sentry.setUser(null);
  }
}

/**
 * Add breadcrumb for debugging
 */
export function addBreadcrumb(
  category: string,
  message: string,
  data?: Record<string, unknown>,
): void {
  if (!sentryInitialized) {
    return;
  }

  Sentry.addBreadcrumb({
    category,
    message,
    data: data ? scrubObject(data) : undefined,
    level: "info",
  });
}

// Export Sentry for direct use if needed
export { Sentry };
